### PR TITLE
SAM-3218: leaving null due date but setting past retract date creates erroneous submisisons

### PR DESCRIPTION
**Pull request by bjones86**
https://jira.sakaiproject.org/browse/SAM-3218

See JIRA for more info and steps to reproduce.

### CODE REVIEW

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SaveAssessmentSettingsListener.java [hunk 1]
```diff
@@ -102,6 +102,13 @@ public void processAction(ActionEvent ae) throws AbortProcessingException
     	context.addMessage(null,new FacesMessage(dueDateErr));
     	error=true;
     }
+
+    if(assessmentSettings.getDueDate() == null && assessmentSettings.getRetractDate() != null && AssessmentAccessControlIfc.ACCEPT_LATE_SUBMISSION.toString().equals(assessmentSettings.getLateHandling())){
+        String dueDateErr = ContextUtil.getLocalizedString("org.sakaiproject.tool.assessment.bundle.AssessmentSettingsMessages", "due_null_with_retract_date");
+        context.addMessage(null,new FacesMessage(dueDateErr));
+        error = true;
+    }
+
```
Comment (ottenhoff) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SaveAssessmentSettingsListener.java]: Can you describe this section a bit? If due date is null and retract date is not null and ACCEPT_LATE_SUBMISSION, then you will set an error and block the user? Is there an action you can take instead of blocking the user?

Comment (bjones86) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SaveAssessmentSettingsListener.java]: No, this is an invalid state. It's a business rule in Samigo that you can't have a late accept date if the due date is null. Allowing this is half of what causes the problem described in the JIRA.

Without this validation error, you could effectively publish the quiz in an invalid state using certain steps (create quiz, give it a due date and a late date, Save+Publish, Edit, null the due date, Save, then Publish). This isn't a new business rule, we're just enforcing it in situations where previously you could circumvent the validation.

Comment (ottenhoff) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SaveAssessmentSettingsListener.java]: So how about just helping the user out a bit instead of throwing up a blocking error: set the due date to the retract date.

Comment (bjones86) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SaveAssessmentSettingsListener.java]: What if they don't want the due date and the retract date to be the same?

Comment (ottenhoff) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SaveAssessmentSettingsListener.java]: Well then they would have set the due date in the first place!

Comment (bjones86) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SaveAssessmentSettingsListener.java]: Well you'd certainly hope so! But I don't think we can make that assumption.

This is the same behavior you would get if you tried to just Save+Publish immediately with a null due date and a non-null retract date; you get the same warning message and the action is blocked. We're just extending the validation to catch this problem when the user takes really weird set of actions to circumvent the exact same validation error message.

Comment (ottenhoff) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SaveAssessmentSettingsListener.java]: If this is for a non-published assessment, then I'm even more confident that we shouldn't be blocking instructor progression and should just do what the vast majority of instructors expect and want. If the instructor truly wants to set separate dates, they have another opportunity when they go through another set of screens to publish this assessment.

Blocking actions are a Sakai peculiarity and are not appreciated by the instructors I have observed. I'm not in favor of adding more blocking actions when we have the opportunity to do what the user expects with a single line of code.

Comment (bjones86) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SaveAssessmentSettingsListener.java]: No, this is for publishing a quiz that has been previously 'Saved'. When saving, none of the date validation is run, because a working copy can have any, null or not. By not publishing right away, but rather saving, then editing to null out the date, then saving, then choosing publish, the user can circumvent the date validation.

If you look at the steps to reproduce in the JIRA, you'll see what I mean in terms of the weird set of steps the user can take to bypass this validation.

Comment (bjones86) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SaveAssessmentSettingsListener.java]: If you try to publish a quiz that has a null due date, but has a retract date set, you will receive the exact same blocking action/message.

Comment (ottenhoff) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SaveAssessmentSettingsListener.java]: > If you try to immediately publish a quiz that has a null due date, but has a retract date set, you will receive the exact same blocking action/message.

So, I'd like to see that changed as well. It's bad behavior. The instructor has a clear goal: stop accepting submissions at XXXX. We shouldn't make this person's life more difficult by setting two dates. Their intent is clear.

Comment (bjones86) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SaveAssessmentSettingsListener.java]: I think if there's a consensus that this validation needs to be removed, then it should be handled in a separate ticket. It will require a lot more work and testing, to ensure all use cases and edge cases are handled properly.

Comment (bjones86) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SaveAssessmentSettingsListener.java]: @ottenhoff, how should we move forward on this?

I think this is a larger issue than what we were initially talking about. Yes, it's a blocking action. However this blocking action was already introduced in a prior JIRA (https://jira.sakaiproject.org/browse/SAM-2985). However the patch on that original issue was lacking protection when the user takes certain actions; this is the blocking action you're referring to on this PR.

Allowing this action also has another negative consequence: allowing the null due date with a valid retract date can give student users the opportunity to take the quiz while it's in an 'invalid' state. This creates erroneous submissions where the student receives 0 for a quiz they shouldn't have been able to take.
